### PR TITLE
Decouple ODEDynamics class from TorchDiffEq

### PR DESCRIPTION
**Pull request by SamWitty**
Addresses #288 .

This refactoring PR introduces a dispatch mechanism for simulating a `ODEDynamics` model with a particular backend passed as a keyword argument to `simulate`, which is then passed to the newly introduced `ode_simulate` function. 

Other to-be-implemented backends now only need to provide implementations for `ode_simulate` and `ode_simulate_to_interruption`.

In addition, this PR introduces a convenience handler `BackendHandler` for inserting the backend argument into encapsulated `simulate` calls.

### CODE REVIEW

**Comment by eb8680** [comment]
There are a bunch of merge conflicts to resolve following #286.

**Comment by SamWitty** [comment]
Strange that there are all of these merge conflicts, as `sw-decouple-ODE` directly branched off of the `sw-dynamics-refactor-2` branch that was used in #286 . I suspect this is something to do with GitHub not recognizing squash-and-merge behavior? Maybe something else? Either way, I'll resolve by accepting all from `sw-decouple-ODE`.

File: chirho/dynamical/internals/ODE/backends/torchdiffeq.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+        torch.transpose(
+            yt_[(..., None) + yt_.shape[len(yt_.shape) - event_dim :]],
+            -len(yt_.shape) - 1,
+            -1 - event_dim,
+        )[0]
+        for yt_ in yt_raw
+    )
+    return yt if event_fn is None else (event_t, yt)
+
+
```
Comment (agrawalraj) [inline chirho/dynamical/internals/ODE/backends/torchdiffeq.py:103]: Code below is new; code above is old

File: chirho/dynamical/internals/ODE/backends/torchdiffeq.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+@ode_simulate.register(TorchDiffEq)
```
Comment (agrawalraj) [inline chirho/dynamical/internals/ODE/backends/torchdiffeq.py:104]: This is the key functionality / small indirection from before

File: chirho/dynamical/ops/ODE/backends/torchdiffeq.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from chirho.dynamical.ops.ODE import ODEBackend
+
+
+class TorchDiffEq(ODEBackend):
```
Comment (agrawalraj) [inline chirho/dynamical/ops/ODE/backends/torchdiffeq.py:4]: This is a class so we can dispatch on it

**Review by agrawalraj — CHANGES_REQUESTED** [review]
Could you also run the notebook to make sure the the inferences still make sense?

**Comment by agrawalraj** [comment]
Otherwise, everything looks really great @SamWitty!

**Review by agrawalraj — APPROVED** [review]
Looks great!

**Pull request merged**